core/storage/wal.rs

```diff
@@ -1108,4 +1364,557 @@ impl WalFileShared {
+                    break;
+                }
+            }
+        }
+    }
+
+    #[test]
+    fn test_wal_read_marks_after_restart() {
+        let (db, _path) = get_database();
+        let wal_shared = db.maybe_shared_wal.read().as_ref().unwrap().clone();
+
+        let conn = db.connect().unwrap();
+        conn.execute("create table test(id integer primary key, value text)")
+            .unwrap();
+        bulk_inserts(&conn, 10, 5);
+        // Checkpoint with restart
+        {
+            let pager = conn.pager.borrow();
+            let mut wal = pager.wal.borrow_mut();
+            let result = run_checkpoint_until_done(&mut *wal, &pager, CheckpointMode::Restart);
+            assert!(result.everything_backfilled());
+        }
+
+        // Verify read marks after restart
+        let read_marks_after: Vec<_> = unsafe {
+            let s = &*wal_shared.get();
+            (0..5)
+                .map(|i| s.read_locks[i].value.load(Ordering::SeqCst))
+                .collect()
+        };
+
+        assert_eq!(read_marks_after[0], 0, "Slot 0 should remain 0");
+        assert_eq!(
+            read_marks_after[1], 0,
+            "Slot 1 (default reader) should be reset to 0"
+        );
+        for (i, item) in read_marks_after.iter().take(5).skip(2).enumerate() {
+            assert_eq!(
+                *item, READMARK_NOT_USED,
+                "Slot {i} should be READMARK_NOT_USED after restart",
+            );
+        }
+    }
+
+    #[test]
+    fn test_wal_concurrent_readers_during_checkpoint() {
+        let (db, _path) = get_database();
+        let conn_writer = db.connect().unwrap();
+
+        conn_writer
+            .execute("create table test(id integer primary key, value text)")
+            .unwrap();
+        bulk_inserts(&conn_writer, 5, 10);
+
+        // Start multiple readers at different points
+        let conn_r1 = db.connect().unwrap();
+        let conn_r2 = db.connect().unwrap();
+
+        // R1 starts reading
+        {
+            let pager = conn_r1.pager.borrow_mut();
+            let mut wal = pager.wal.borrow_mut();
+            assert!(matches!(wal.begin_read_tx().unwrap(), LimboResult::Ok));
+        }
+        bulk_inserts(&conn_writer, 5, 10);
+
+        // R2 starts reading, sees more frames than R1
+        let r2_max_frame = {
+            let pager = conn_r2.pager.borrow_mut();
+            let mut wal = pager.wal.borrow_mut();
+            assert!(matches!(wal.begin_read_tx().unwrap(), LimboResult::Ok));
+            wal.get_max_frame()
+        };
+
+        // try passive checkpoint, should only checkpoint up to R1's position
+        let checkpoint_result = {
+            let pager = conn_writer.pager.borrow();
+            let mut wal = pager.wal.borrow_mut();
+            run_checkpoint_until_done(&mut *wal, &pager, CheckpointMode::Passive)
+        };
+
+        assert!(
+            checkpoint_result.num_checkpointed_frames < checkpoint_result.num_wal_frames,
```

Comment:
Can we make this assertion more precise (i.e. we know we only checkpointed up to r1's mark)?